src/tokio/io.rs

```diff
@@ -381,7 +381,7 @@ where
 
                 poll
             }
-            Status::Disconnected(_) => Poll::Pending,
+            Status::Disconnected(_) => exhausted_err(),
```

Comment:
I think this should be Ok and not an error.

```rs
Status::Disconnected(_) => Poll::Ready(Ok(())),
```

In your test, I replaced stubborn-io::TcpStream with the tokio::net::TcpStream implementation, and the behavior I observed is that it returns an Ok, not an error, so we should try to match it. You were right to change it away from pending.

Reply:
Interesting, perhaps this is platform dependent? I get a panic running the below test on the second call to shutdown

```rust
#[tokio::test]
async fn back_to_back_shutdown_attempts() {
    const ADDR: &str = "127.0.0.1:3989";

    tokio::spawn(async move {
        let mut streams = Vec::new();
        let listener = tokio::net::TcpListener::bind(ADDR).await.unwrap();
        loop {
            let (stream, _addr) = listener.accept().await.unwrap();
            streams.push(stream);
        }
    });
    tokio::time::sleep(Duration::from_secs(1)).await;
    let mut connection = TcpStream::connect(ADDR).await.unwrap();

    connection.shutdown().await.unwrap();
    connection.shutdown().await.unwrap();
}
```

```
called `Result::unwrap()` on an `Err` value: Os { code: 57, kind: NotConnected, message: "Socket is not connected" }
```

Reply:
Are you on Windows? I am running this on Ubuntu and I keep seeing the non error. If you are on Windows, maybe this inconsistency is a bug in tokio. I agree with you that this seems like disconnecting from a connection should be an error if it is already disconnected.

Reply:
Found this in std::net::tcp docstrings:

```
Shuts down the read, write, or both halves of this connection.
This function will cause all pending and future I/O on the specified portions to return immediately with an appropriate value (see the documentation of Shutdown).
Platform-specific behavior
Calling this function multiple times may result in different behavior, depending on the operating system. On Linux, the second call will return Ok(()), but on macOS, it will return ErrorKind::NotConnected. This may change in the future.
```

Reply:
Maybe it makes sense returning an error, but instead of returning the exhausted_err, you can return something like an already_shutdown error or something. What do you think?

Reply:
Oh very interesting, I did not know that. I guess that explains it, I'm on macOS.

I would say that while in this example I'm calling `AsyncWrite::shutdown` twice to trigger this behavior, the issue is maybe slightly broader. We should craft the error so that it's general enough for any call to `AsyncWrite::poll_shutdown` where the `UnderlyingIo` is in the disconnected state. Another example I could think of is calling `tokio::io::split`. Perhaps in one thread the reader triggers a disconnect, and in a second we unknowingly call shutdown on the writer. So ideally an error which captures the fact that the I/O is currently disconnected.

Reply:
If you want, please add the appropriate error, and also, bump the patch version in the cargo.toml.

Thank you!

Reply:
Added commit with new error message and bumped version in Cargo.toml.